include/votca/xtp/bseoperator_btda.h

```diff
@@ -0,0 +1,232 @@
+    }
+    return diag;
+  }
+
+  Eigen::VectorXd diagonal() const {return _diag;}
+  
+  // get the full matrix if we have to
+  Eigen::MatrixXd get_full_matrix() const {
+    Eigen::MatrixXd matrix = Eigen::MatrixXd::Zero(_size, _size);
+  
+    for (int i = 0; i < _size; i++) {
+      matrix.row(i) = row(i);
+    }
+    return matrix;
+  }
+
+ private:
+
+  const MatrixReplacementA & _A;
+  const MatrixReplacementB & _B;
+  int _size;
+  Eigen::VectorXd _diag; 
+
+};
+}  // namespace xtp
+}  // namespace votca
+
+namespace Eigen {
+namespace internal {
+
+// replacement of the mat*vect operation
+template <typename Vtype, typename MatrixReplacementA, typename MatrixReplacementB>
+struct generic_product_impl<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Vtype, DenseShape,
+                            DenseShape, GemvProduct>
+    : generic_product_impl_base<
+          votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Vtype,
+          generic_product_impl<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Vtype> >{
+
+  typedef
+      typename Product<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Vtype>::Scalar Scalar;
+
+  template <typename Dest>
+  static void scaleAndAddTo(Dest& dst, const votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>& op,
+                            const Vtype& v, const Scalar& alpha) {
+    // returns dst = alpha * op * v
+    // alpha must be 1 here
+    assert(alpha == Scalar(1) && "scaling is not implemented");
+    EIGEN_ONLY_USED_FOR_DEBUG(alpha);
+
+// make the mat vect product
+#pragma omp parallel for
+    for (int i = 0; i < op.rows(); i++) {
+      dst(i) = op.row(i) * v;
+    }
+  }
+};
+
+// replacement of the mat*mat operation
+template <typename Mtype, typename MatrixReplacementA, typename MatrixReplacementB>
+struct generic_product_impl<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, 
+                            Mtype, DenseShape, DenseShape, GemmProduct>
+    : generic_product_impl_base<
+          votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Mtype,
+          generic_product_impl<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Mtype>> {
+
+  typedef
+      typename Product<votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>, Mtype>::Scalar Scalar;
+
+  template <typename Dest>
+  static void scaleAndAddTo(Dest& dst, const votca::xtp::HamiltonianOperator<MatrixReplacementA,MatrixReplacementB>& op,
+                            const Mtype& m, const Scalar& alpha) {
+    // returns dst = alpha * op * v
+    // alpha must be 1 here
+    assert(alpha == Scalar(1) && "scaling is not implemented");
+    EIGEN_ONLY_USED_FOR_DEBUG(alpha);
+
+// make the mat mat product
+#pragma omp parallel for
+    for (int i = 0; i < op.rows(); i++) {
+      const Eigen::RowVectorXd row = op.row(i) * m;
```

Comment:
as above

Reply:
done. Note that _A and _B must be public then ...